src/Functions/substringIndex.cpp

```diff
@@ -0,0 +1,309 @@
+                    "Illegal type {} of second argument of function {}",
+                    arguments[1]->getName(),
+                    getName());
+
+            if (!isNativeNumber(arguments[2]))
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Illegal type {} of third argument of function {}",
+                    arguments[2]->getName(),
+                    getName());
+
+            return std::make_shared<DataTypeString>();
+        }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t /*input_rows_count*/) const override
+        {
+            ColumnPtr column_string = arguments[0].column;
+            ColumnPtr column_delim = arguments[1].column;
+            ColumnPtr column_index = arguments[2].column;
+
+            const ColumnConst * column_delim_const = checkAndGetColumnConst<ColumnString>(column_delim.get());
+            if (!column_delim_const)
+                throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Second argument to {} must be a constant String", getName());
+
+            String delim = column_delim_const->getValue<String>();
+            if constexpr (!is_utf8)
+            {
+                if (delim.size() != 1)
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Second argument to {} must be a single character", getName());
+            }
+            else
+            {
+                if (UTF8::countCodePoints(reinterpret_cast<const UInt8 *>(delim.data()), delim.size()) != 1)
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Second argument to {} must be a single UTF-8 character", getName());
+            }
+
+            auto column_res = ColumnString::create();
+            ColumnString::Chars & vec_res = column_res->getChars();
+            ColumnString::Offsets & offsets_res = column_res->getOffsets();
+
+            const ColumnConst * column_string_const = checkAndGetColumnConst<ColumnString>(column_string.get());
+            if (column_string_const)
+            {
+                String str = column_string_const->getValue<String>();
+                constantVector(str, delim, column_index.get(), vec_res, offsets_res);
+            }
+            else
+            {
+                const auto * col_str = checkAndGetColumn<ColumnString>(column_string.get());
+                if (!col_str)
+                    throw Exception(ErrorCodes::ILLEGAL_COLUMN, "First argument to {} must be a String", getName());
+
+                bool is_index_const = isColumnConst(*column_index);
+                if (is_index_const)
+                {
+                    Int64 index = column_index->getInt(0);
+                    vectorConstant(col_str, delim, index, vec_res, offsets_res);
+                }
+                else
+                    vectorVector(col_str, delim, column_index.get(), vec_res, offsets_res);
+            }
+            return column_res;
```

Comment:
constantConstant is impossible because useDefaultImplementationForConstants returns true.

Reply:
> maybe early returns instead of nested ifs would be better... and what about `constantConstant`?..

It depends, in this case if else branchs do not handle exceptions, they have the same status.